File: src/infra/captcha_service/stub.rs
use super::{CaptchaService, ValidateError};
use crate::domain::models::Captcha;
use async_trait::async_trait;

pub struct StubCaptchaService {
    error: bool,
}

impl StubCaptchaService {
    #[cfg(test)]
    pub fn new() -> Self {
        Self { error: false }
    }

    #[cfg(test)]
    pub fn with_error(self) -> Self {
        Self {
            error: true,
            ..self
        }
    }
}

#[async_trait]
impl CaptchaService for StubCaptchaService {
    async fn validate(&self, _captcha: Captcha) -> Result<(), ValidateError> {
        if self.error {
            Err(ValidateError::Unknown("error flag is on"))
        } else {
            Ok(())
        }
    }
}

File: src/infra/captchaService/stub.ts
import { CaptchaService } from "./mod.ts";
import { Captcha } from "@domain";
import { Result } from "@utils";

export class StubCaptchaService implements CaptchaService {
    private error: boolean;

    private constructor(error: boolean) {
        this.error = error;
    }

    static new() {
        return new StubCaptchaService(false);        
    }

    static withError() {
        return new StubCaptchaService(true);
    }

    validate(_: Captcha): Promise<Result<void, string>> {
        if (this.error) {
            return Promise.resolve(Result.err("error flag is on"));
        } else {
            return Promise.resolve(Result.ok(undefined));
        }
    }
};
